### PR TITLE
W3 - D1 - BaoLe - Android - File Storage

### PR DESCRIPTION
**Pull request by at-baole**
W3 - D1 - BaoLe - Android - File Storage #21

### CODE REVIEW

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            } catch (IOException ignored) {
+            }
+        } else {
+            Toast.makeText(getActivity(), R.string.inputNull, Toast.LENGTH_LONG).show();
+        }
+    }
+
+    public void loadDataInternal() {
+        try (FileInputStream fis = Objects.requireNonNull(getActivity()).openFileInput(FILE_NAME)) {
+            InputStreamReader isr = new InputStreamReader(fis);
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.filestorage.InternalExternalFragment.loadDataInternal(): new java.io.InputStreamReader(InputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        mBtnSaveInternal.setOnClickListener(intSaveView -> saveDataInternal());
+        mBtnLoadInternal.setOnClickListener(intLoadView -> loadDataInternal());
+        return view;
+    }
+
+    public void saveDataInternal() {
+        String text = mEdtInputInternal.getText().toString();
+        if (mEdtInputInternal.getText().length() != 0) {
+            try (FileOutputStream fos = Objects.requireNonNull(getActivity()).openFileOutput(FILE_NAME, Context.MODE_PRIVATE)) {
+                fos.write(text.getBytes());
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.filestorage.InternalExternalFragment.saveDataInternal(): String.getBytes()
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-baole/res/layout/activity_main.xml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:gravity="center"
+        android:text="@string/listIssues"
+        android:textSize="30sp"
+        android:textColor="@color/colorRed"/>
+
+    <Button
+        android:id="@+id/btnToGroupView"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToGroupView" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:layout_marginTop="10dp"
+        android:text="@string/toViewAndGroupView"/>
+
+    <Button
+        android:id="@+id/btnToEventListener"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToEventListener" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:layout_marginTop="10dp"
+        android:text="@string/toEventListener"/>
+
+    <Button
+        android:id="@+id/btnToActivityFragment"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToActivityFragment" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:layout_marginTop="10dp"
+        android:text="@string/toActivityFragment"/>
+
+    <Button
+        android:id="@+id/btnToRecyclerView"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToRecyclerView" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:layout_marginTop="10dp"
+        android:text="@string/toRecyclerView"/>
+
+    <Button
+        android:id="@+id/btnToFileStorage"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToFileStorage" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    private void loadDataExternal() {
+        File Root = Environment.getExternalStorageDirectory();
+        File Dir = new File(Root.getAbsolutePath() + "/MyAppFile");
+        File file = new File(Dir, "MyMessage.txt");
+        String Message;
+        try {
+            FileInputStream fis = new FileInputStream(file);
+            InputStreamReader isr = new InputStreamReader(fis);
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.filestorage.InternalExternalFragment.loadDataExternal(): new java.io.InputStreamReader(InputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            File Root = Environment.getExternalStorageDirectory();
+            File Dir = new File(Root.getAbsolutePath() + "/MyAppFile");
+            if (!Dir.exists()) {
+                Dir.mkdir();
+            }
+            File file = new File(Dir, "MyMessage.txt");
+            String Message = mEdtInputExternal.getText().toString();
+            try {
+                FileOutputStream fos = new FileOutputStream(file);
+                fos.write(Message.getBytes());
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.filestorage.InternalExternalFragment.saveDataExternal(): String.getBytes()
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-baole/AndroidManifest.xml [hunk 10]
```diff
@@ -10,16 +10,23 @@
...
-        <activity android:name=".activity_fragment.LoginActivity"/>
+        <activity android:name=".view_and_groupview.ViewActivity" />
+        <activity android:name=".activity_fragment.InformationActivity" />
+        <activity android:name=".activity_fragment.LoginActivity" />
+        <activity android:name=".recyclerview.RecyclerViewActivity" />
+        <activity android:name=".filestorage.FileStorageActivity" />
+
     </application>
+
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```
Comment (at-dev-support-01) [inline app/src/at-baole/AndroidManifest.xml]: WARNING: [ManifestOrder] `<uses-permission>` tag appears after `<application>` tag
 The <application> tag should appear after the elements which declare which version you need, which features you need, which libraries you need, and so on. In the past there have been subtle bugs (such as themes not getting applied correctly) when the `<application>` tag appears before some of these other elements, so it's best to order your manifest in the logical dependency order.

File: app/src/at-baole/res/layout/activity_employee.xml [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    </LinearLayout>
+
+    <TextView
+        android:layout_marginTop="20dp"
+        android:layout_width="match_parent"
+        android:layout_height="40dp"
+        android:gravity="center"
+        android:textColor="@color/colorRed"
+        android:background="@drawable/border"
+        android:text="LIST EMPLOYEE"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_employee.xml]: WARNING: [HardcodedText] Hardcoded string "LIST EMPLOYEE", should use `@string` resource
 Hardcoding text attributes directly in layout files is bad for several reasons:

* When creating configuration variations (for example for landscape or portrait) you have to repeat the actual text (and keep it up to date when making changes)

* The application cannot be translated to other languages by just adding new translations for existing string resources.

There are quickfixes to automatically extract this hardcoded string into a resource lookup.

File: app/src/at-baole/res/layout/activity_employee.xml [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="40dp">
+
+        <TextView
+            android:layout_width="0dp"
+            android:layout_height="match_parent"
+            android:layout_weight="3"
+            android:background="@drawable/border"
+            android:gravity="center"
+            android:text="Emp. ID"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_employee.xml]: WARNING: [HardcodedText] Hardcoded string "Emp. ID", should use `@string` resource
 Hardcoding text attributes directly in layout files is bad for several reasons:

* When creating configuration variations (for example for landscape or portrait) you have to repeat the actual text (and keep it up to date when making changes)

* The application cannot be translated to other languages by just adding new translations for existing string resources.

There are quickfixes to automatically extract this hardcoded string into a resource lookup.

File: app/src/at-baole/res/layout/activity_employee.xml [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            android:textSize="20sp" />
+
+        <TextView
+            android:layout_width="0dp"
+            android:layout_height="match_parent"
+            android:layout_weight="7"
+            android:background="@drawable/border"
+            android:gravity="center"
+            android:text="Employee Name"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_employee.xml]: WARNING: [HardcodedText] Hardcoded string "Employee Name", should use `@string` resource
 Hardcoding text attributes directly in layout files is bad for several reasons:

* When creating configuration variations (for example for landscape or portrait) you have to repeat the actual text (and keep it up to date when making changes)

* The application cannot be translated to other languages by just adding new translations for existing string resources.

There are quickfixes to automatically extract this hardcoded string into a resource lookup.

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    private void loadDataExternal() {
+        File Root = Environment.getExternalStorageDirectory();
+        File Dir = new File(Root.getAbsolutePath() + "/" + EXTERNAL_DIRECTORY_NAME);
+        File file = new File(Dir, EXTERNAL_FILE_NAME);
+        String input;
+        try {
+            FileInputStream fis = new FileInputStream(file);
+            InputStreamReader isr = new InputStreamReader(fis);
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.filestorage.InternalExternalFragment.loadDataExternal(): new java.io.InputStreamReader(InputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-baole/res/values/colors.xml [hunk 15]
```diff
@@ -11,5 +11,5 @@
     <color name="colorGrey">#b3b3b3</color>
     <color name="colorLightGrey">#e6e6e6</color>
     <color name="colorLightBlue">#9999ff</color>
-    <color name="colorCyan">#33cccc</color>
+    <color name="colorRed">#ff0000</color>
 </resources>
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/values/colors.xml]: ERROR: [MergeMarker] Missing merge marker?
 Many version control systems leave unmerged files with markers such as <<< in the source code. This check looks for these markers, which are sometimes accidentally left in, particularly in resource files where they don't break compilation.

File: app/src/at-baole/res/layout/activity_main.xml (hunk 3 above)
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToGroupView" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml (hunk 4 above)
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToEventListener" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml (hunk 5 above)
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToActivityFragment" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml (hunk 6 above)
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToRecyclerView" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/res/layout/activity_main.xml (hunk 7 above)
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToFileStorage" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
+package asiantech.internship.summer.mainactivity;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.v7.app.AppCompatActivity;
+import android.view.View;
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.view.View'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+import android.widget.Button;
+
+import asiantech.internship.summer.R;
+import asiantech.internship.summer.activity_fragment.LoginActivity;
+import asiantech.internship.summer.event_listener.SignUpActivity;
+import asiantech.internship.summer.filestorage.FileStorageActivity;
+import asiantech.internship.summer.filestorage.InternalExternalFragment;
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'asiantech.internship.summer.filestorage.InternalExternalFragment'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+import asiantech.internship.summer.recyclerview.RecyclerViewActivity;
+import asiantech.internship.summer.view_and_groupview.ViewActivity;
+import asiantech.internship.summer.viewpager.PagerActivity;
+import asiantech.internship.summer.viewpager.TimelineActivity;
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/mainactivity/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'asiantech.internship.summer.viewpager.TimelineActivity'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-baole/res/layout/activity_main.xml [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+        style="@style/Button"
+        android:text="@string/toActivityFragment" />
+
+    <Button
+        android:id="@+id/btnToRecyclerView"
+        style="@style/Button"
+        android:text="@string/toRecyclerView" />
+
+    <Button
+        android:id="@+id/btnToViewPager"
```
Comment (at-dev-support-01) [inline app/src/at-baole/res/layout/activity_main.xml]: WARNING: [InconsistentLayout] The id "btnToViewPager" in layout "activity_main" is missing from the following layout configurations:  (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There **are** cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/SharedPreferencesFragment.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+    private EditText mEdtUsername;
+    private EditText mEdtPassword;
+    private Button mBtnLogin;
+    private SharedPreferences mPreferences;
+    private SharedPreferences.Editor mEditor;
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/SharedPreferencesFragment.java]: Please move savedInstanceState param  to line 26.

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    private static final String EXTERNAL_FILE_NAME = "externalExample.txt";
+    private static final String EXTERNAL_DIRECTORY_NAME = "MyAppFile";
+    private static final int REQUEST_ID_WRITE_PERMISSION = 200;
+    private static final int SDK_VERSION_CHECK = 23;
+    private EditText mEdtInputInternal;
+    private EditText mEdtInputExternal;
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: Move savedInstanceState param to line 37.

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+    private void saveDataExternal() {
+        String state;
+        state = Environment.getExternalStorageState();
+        if (Environment.MEDIA_MOUNTED.equals(state)) {
+            File Root = Environment.getExternalStorageDirectory();
+            File Dir = new File(Root.getAbsolutePath() + "/" + EXTERNAL_DIRECTORY_NAME);
+            if (!Dir.exists()) {
+                Dir.mkdir();
+            }
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: Please remove it. it is unnecessary.
```
if (!Dir.exists()) {
	 Dir.mkdir();
 }
```

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+                sb.append(input).append("\n");
+            }
+            mEdtInputExternal.setText(sb.toString());
+        } catch (IOException ignored) {
+        }
+    }
+
+    private void askPermissionAndWriteFile() {
+        boolean canWrite = this.askPermission();
+        //
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: Please remove it.

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/DataAccess.java [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+        cursor.close();
+        db.close();
+        return companies;
+    }
+
+    int getCompaniesCount() {
+        String countQuery = "SELECT * FROM " + TABLE_COMPANY;
+        SQLiteDatabase db = this.getReadableDatabase();
+        Cursor cursor = db.rawQuery(countQuery, null);
+        int countCompaniew = cursor.getCount();
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/DataAccess.java]: ``` countCompaniew ``` rename to ```count```

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/SaveDatabaseFragment.java [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+import asiantech.internship.summer.R;
+import asiantech.internship.summer.filestorage.model.Company;
+
+public class SaveDatabaseFragment extends Fragment implements CompanyAdapter.OnClickCompany {
+    public static final String ID_COMPANY = "CompanyId";
+    DataAccess mDataAccess;
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/SaveDatabaseFragment.java]: Please add modifier.

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/SaveDatabaseFragment.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+            addListCompany();
+        }
+        initCompany(view);
+        return view;
+    }
+
+    private void initCompany(View view) {
+        RecyclerView mRecyclerView = view.findViewById(R.id.recyclerViewCompany);
+        mRecyclerView.setHasFixedSize(true);
+        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/SaveDatabaseFragment.java]: change to
 LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/EmployeeActivity.java [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+        Intent intent = getIntent();
+        mCompanyId = intent.getIntExtra(SaveDatabaseFragment.ID_COMPANY, 0);
+        initEmployeeView();
+    }
+
+    private void initEmployeeView() {
+        Company mCompany = mDataAccess.getCompanyById(mCompanyId);
+        RecyclerView mRvEmployee = findViewById(R.id.rvEmployee);
+        TextView mTvNameCompany = findViewById(R.id.tvCompany);
+        Button mBtnInsert = findViewById(R.id.btnInsert);
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/EmployeeActivity.java]: Please remove prefix ```m```

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/EmployeeActivity.java [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+        mBtnUpdate = findViewById(R.id.btnUpdate);
+        mEdtEmployeeId = findViewById(R.id.edtIdEmployee);
+        mEdtEmployeeName = findViewById(R.id.edtEmployeeName);
+
+        mBtnInsert.setOnClickListener(this);
+        mBtnDelete.setOnClickListener(this);
+        mBtnUpdate.setOnClickListener(this);
+
+        mRvEmployee.setHasFixedSize(true);
+        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/EmployeeActivity.java]: Please use
 LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this); stead of

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+            Toast.makeText(getActivity(), R.string.inputNull, Toast.LENGTH_LONG).show();
+        }
+    }
+
+    public void loadDataInternal() {
+        try (FileInputStream fis = Objects.requireNonNull(getActivity()).openFileInput(INTERNAL_FILE_NAME)) {
+            InputStreamReader isr = new InputStreamReader(fis, getString(R.string.encoding));
+            BufferedReader br = new BufferedReader(isr);
+            StringBuilder sb;
+            sb = new StringBuilder();
```
Comment (at-chauhoang) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/InternalExternalFragment.java]: Please place same line
 StringBuilder sb = new StringBuilder();

File: app/src/at-baole/java/asiantech/internship/summer/filestorage/FileStorageActivity.java [hunk 30]
```diff
@@ -0,0 +1,31 @@
+package asiantech.internship.summer.filestorage;
+
+import android.os.Bundle;
+import android.support.design.widget.TabLayout;
+import android.support.v4.app.FragmentTransaction;
```
Comment (at-dev-support-01) [inline app/src/at-baole/java/asiantech/internship/summer/filestorage/FileStorageActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.support.v4.app.FragmentTransaction'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

**Comment by at-baole** [comment]
Thank you so much! I fixed comments. Please review again.

**Pull request merged**